doc/specs/stdlib_string_type.md

```diff
@@ -0,0 +1,1607 @@
+---
+title: string type
+---
+
+# The `stdlib_string_type` module
+
+[TOC]
+
+## Introduction
+
+The `stdlib_string_type` provides a derived type holding an arbitrary sequence
+of characters compatible with most Fortran intrinsic character procedures as
+well as operators for working with character variables and constants.
+
+
+## Derived types provided
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### The `string_type` derived type
+
+The `string_type` is defined as a non-extenable derived type representing a
+sequence of characters. The internal representation of the character sequence
+is implementation dependent and not visible for the user of the module.
+
+#### Status
+
+Experimental
+
+
+## Procedures and methods provided
+
+Procedures returning `string_type` instances can usually used in elemental
+context, while procedures returning scalar character values can only be
+used in a pure way.
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Constructor for empty string
```

Comment:
The two constructors could be combine into one single section, as:

`res = [[stdlib_string_type(module):string_type(interface)]] ([string])`

with `string` being `optional`

Reply:
That's how it is implemented, but for the API I found it clearer to specify the construction from a deferred length character separately from the empty string construction.

Reply:
It was first a bit confusing to me. Let see what others think about it.

Reply:
Perhaps a more philosophical question, but if it is clearer this way in specs, why not also implement two constructors with a generic interface? (Answer is obviously to avoid code duplication).

Doesn't the non-fancy functional string behave as an empty string by default if not explicitly initialized? Are there any situations where I might want to explicitly use `string_type()` instead of just `''`?

Reply:
I think there is an issue with two separate constructors, the only way to implement the specs is using the allocatable to optional mapping.

While I like this mapping it might not be apparent for somebody not too familiar with Fortran. Therefore, I thought explaining the resulting cases as separate constructors is clearer. It depends on whether specs/docs here are targeted at implementors or users whatever to prefer.

If you have declared it as variable, yes it does. In case you need to pass a string to a function that only accepts `string_type` instances the empty constructor becomes handy.

Reply:
> I think there is an issue with two separate constructors, the only way to implement the specs is using the allocatable to optional mapping.

Is this due to the choice of deferred-length character for the character storage sequence?

From my point of view, it is okay two keep the two constructor forms separate in the specs.

Reply:
> Is this due to the choice of deferred-length character for the character storage sequence?

Not really, just the usual Fortran madness about deferred length character vs fixed length character handling, sigh.